Project 2/k-NNAlgorithms/src/Algorithms.java

```diff
@@ -0,0 +1,171 @@
+
+
+            //find average if using regression data
+            if(regression){
+                results.add(MathFunction.average(classificationOfNeighbors));
+            //find mode with classifcation data
+            }else{
+                results.add(MathFunction.mode(classificationOfNeighbors));
+            }
+
+        }
+        //returns the list of guessed results for the test set
+        return results;
+    }
+
+
+
+    public static ArrayList<String> EditedKNN(ArrayList<ArrayList<String>> trainingData, ArrayList<ArrayList<String>> testingData, ArrayList<ArrayList<String>>validationSet, int k, boolean regression, boolean euclidean) {
+
+        int lengthOfTrainingData= trainingData.size();
+        int indexOfClassification = trainingData.get(0).size()-1;
+        double currentPrecision = 0.00000001;
+        double currentRecall=0.0000001;
+        double currentAccuracy=.0000001;
+        double prevPrecision = 0;
+        double prevRecall=0;
+        double prevAccuracy=0;
+
+        ArrayList<Integer> badSamplePoints= new ArrayList<>();//arraylist to hold the indexes of the points that were incorrectly classified
+
+        ArrayList<ArrayList<String>> editedTrainingData =(ArrayList<ArrayList<String>>) trainingData.clone();//holds the edited training data
+        ArrayList<ArrayList<String>> finalEditedTrainingData =new ArrayList<>();
+
+
+        while(currentAccuracy>prevAccuracy && currentPrecision>prevPrecision && currentRecall>prevRecall){//while edited set is still improving
+
+            finalEditedTrainingData=(ArrayList<ArrayList<String>>) editedTrainingData.clone();//copy the editedData into a potential final set
+            for (int i = 0; i < editedTrainingData.size(); i++) {//go through each point in the editedData
+                ArrayList<ArrayList<String>> samplePoint = new ArrayList<>();//make a new sample point
+                samplePoint.add(editedTrainingData.get(i));//give it the value of editedTraining data at i
+                ArrayList<String> classification = KNN(editedTrainingData,samplePoint, k, regression,euclidean);
+
+                if(classification.get(0).equals(samplePoint.get(0).get(indexOfClassification))){ //sample point classified correctly
+                    //do nothing
+                }else{//sample point not correctly classified
+                    editedTrainingData.remove(editedTrainingData.get(i));//remove bad data from the set
+                }
+            }
+
+            //run the editedTraining set with the validation set to determine if
+            //edited increased or decreased accuracy, precision or recall
+            ArrayList<String> results= MathFunction.processConfusionMatrix(KNN(editedTrainingData,validationSet,k,regression,euclidean), validationSet);
+
+            //set prevs and currs to determine if a decrease in accuracy,precision or recall has
+            //occurred, exit while loop if so.
+            prevAccuracy= currentAccuracy+.05;
```

Comment:
Just a note that we should tune this and comment why we're adding to this additional factor.